packages/flutter/lib/src/cupertino/date_picker.dart

```diff
@@ -11,10 +11,707 @@ import 'picker.dart';
+
+      amPmController = new FixedExtentScrollController(initialItem: selectedAmPm);
+    }
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+
+    textDirectionFactor = Directionality.of(context) == TextDirection.ltr ? 1 : -1;
+    localizations = CupertinoLocalizations.of(context) ?? const DefaultCupertinoLocalizations();
+
+    alignCenterLeft = textDirectionFactor == 1 ? Alignment.centerLeft : Alignment.centerRight;
+    alignCenterRight = textDirectionFactor == 1 ? Alignment.centerRight : Alignment.centerLeft;
+  }
+
+  // Gets the current date time of the picker.
+  DateTime _getDateTime() {
+    final DateTime date = new DateTime(
+      widget.initialDateTime.year,
+      widget.initialDateTime.month,
+      widget.initialDateTime.day,
+    ).add(Duration(days: selectedDayDifference));
+
+    return new DateTime(
+      date.year,
+      date.month,
+      date.day,
+      selectedHour + selectedAmPm * 12,
+      selectedMinute,
+    );
+  }
+
+  Widget _buildMediumDatePicker(double offAxisFraction, Function childPositioning) {
+    return new CupertinoPicker.builder(
+      scrollController: new FixedExtentScrollController(initialItem: selectedDayDifference),
+      offAxisFraction: offAxisFraction,
+      itemExtent: _kItemExtent,
+      useMagnifier: _kUseMagnifier,
+      magnification: _kMagnification,
+      backgroundColor: _kBackgroundColor,
+      onSelectedItemChanged: (int index) {
+        selectedDayDifference = index;
+        widget.onDateTimeChanged(_getDateTime());
+      },
+      itemBuilder: (BuildContext context, int index) {
+        final DateTime dateTime = new DateTime(
+          widget.initialDateTime.year,
+          widget.initialDateTime.month,
+          widget.initialDateTime.day,
+        ).add(Duration(days: index));
+
+        if (widget.minimumDate != null && dateTime.isBefore(widget.minimumDate))
+          return null;
+        if (widget.maximumDate != null && dateTime.isAfter(widget.maximumDate))
+          return null;
+
+        return childPositioning(Text(localizations.datePickerMediumDate(dateTime)));
+      },
+    );
+  }
+
+  Widget _buildHourPicker(double offAxisFraction, Function childPositioning) {
+    return new CupertinoPicker(
+      scrollController: new FixedExtentScrollController(initialItem: selectedHour),
+      offAxisFraction: offAxisFraction,
+      itemExtent: _kItemExtent,
+      useMagnifier: _kUseMagnifier,
+      magnification: _kMagnification,
+      backgroundColor: _kBackgroundColor,
+      onSelectedItemChanged: (int index) {
+        if (widget.use24hFormat) {
+          selectedHour = index;
+          widget.onDateTimeChanged(_getDateTime());
+        }
+        else {
+          // Automatically scrolls the am/pm column if condition is met.
+          if (selectedHour == 0 && index % 12 == 11 || selectedHour == 11 && index % 12 == 0) {
+            selectedHour = index % 12;
+            amPmController.animateToItem(
```

Comment:
By trying manually, it seems like I would fail this kind of test. The animation has to finished before the controller moves to a new one.
Another issue is when scrolling really fast, the `onSelectedItemChanged` function is not called for consecutive items, so the condition here is not satisfied, although we do want to scroll in such condition.
I will think about it. Any suggestion/hint on how to improve this would be great.